File: modules/compiler/python/zap/parser.py
"""Recursive descent parser for Zap. Produces a typed AST."""

from __future__ import annotations
from .ast_nodes import *
from .lexer import Token, TT, tokenize


class ParseError(Exception):
    pass


class Parser:
    def __init__(self, tokens: list[Token]) -> None:
        self._t = tokens
        self._pos = 0

    # ── helpers ───────────────────────────────────────────────────────────────

    def _cur(self) -> Token:
        return self._t[self._pos]

    def _peek(self) -> Token:
        idx = min(self._pos + 1, len(self._t) - 1)
        return self._t[idx]

    def _at(self, *types: TT) -> bool:
        return self._cur().ty in types

    def _advance(self) -> Token:
        tok = self._t[self._pos]
        self._pos += 1
        return tok

    def _eat(self, ty: TT) -> Token:
        tok = self._cur()
        if tok.ty != ty:
            raise ParseError(
                f"Line {tok.line}: expected {ty.name}, got {tok.ty.name} ({tok.value!r})"
            )
        return self._advance()

    # ── top level ─────────────────────────────────────────────────────────────

    def parse_program(self) -> Program:
        decls: list[FnDecl | Stmt] = []
        while not self._at(TT.EOF):
            if self._at(TT.FN):
                decls.append(self._fn())
            else:
                decls.append(self._stmt())
        return Program(decls)

    def _fn(self) -> FnDecl:
        line = self._cur().line
        self._eat(TT.FN)
        name = self._eat(TT.IDENT).value
        self._eat(TT.LPAREN)
        params: list[tuple[str, str]] = []
        if not self._at(TT.RPAREN):
            params.append(self._param())
            while self._at(TT.COMMA):
                self._advance()
                params.append(self._param())
        self._eat(TT.RPAREN)
        self._eat(TT.ARROW)
        ret = self._type()
        body = self._block()
        return FnDecl(name, params, ret, body, line)

    def _param(self) -> tuple[str, str]:
        name = self._eat(TT.IDENT).value
        self._eat(TT.COLON)
        return (name, self._type())

    def _type(self) -> str:
        tok = self._cur()
        if tok.ty == TT.T_INT:  self._advance(); return "int"
        if tok.ty == TT.T_BOOL: self._advance(); return "bool"
        if tok.ty == TT.T_VOID: self._advance(); return "void"
        raise ParseError(f"Line {tok.line}: expected type, got {tok.value!r}")

    def _block(self) -> Block:
        self._eat(TT.LBRACE)
        stmts: list[Stmt] = []
        while not self._at(TT.RBRACE, TT.EOF):
            stmts.append(self._stmt())
        self._eat(TT.RBRACE)
        return Block(stmts)

    # ── statements ────────────────────────────────────────────────────────────

    def _stmt(self) -> Stmt:
        match self._cur().ty:
            case TT.LET:    return self._let()
            case TT.RETURN: return self._return()
            case TT.IF:     return self._if()
            case TT.WHILE:  return self._while()
            case TT.PRINT:  return self._print()
            case _:         return self._expr_stmt()

    def _let(self) -> LetStmt:
        line = self._cur().line
        self._eat(TT.LET)
        name = self._eat(TT.IDENT).value
        self._eat(TT.COLON)
        ty = self._type()
        self._eat(TT.EQ)
        val = self._expr()
        self._eat(TT.SEMI)
        return LetStmt(name, ty, val, line)

    def _return(self) -> ReturnStmt:
        line = self._cur().line
        self._eat(TT.RETURN)
        val = None if self._at(TT.SEMI) else self._expr()
        self._eat(TT.SEMI)
        return ReturnStmt(val, line)

    def _if(self) -> IfStmt:
        line = self._cur().line
        self._eat(TT.IF)
        cond = self._expr()
        then = self._block()
        else_ = None
        if self._at(TT.ELSE):
            self._advance()
            else_ = self._block()
        return IfStmt(cond, then, else_, line)

    def _while(self) -> WhileStmt:
        line = self._cur().line
        self._eat(TT.WHILE)
        cond = self._expr()
        body = self._block()
        return WhileStmt(cond, body, line)

    def _print(self) -> PrintStmt:
        line = self._cur().line
        self._eat(TT.PRINT)
        self._eat(TT.LPAREN)
        val = self._expr()
        self._eat(TT.RPAREN)
        self._eat(TT.SEMI)
        return PrintStmt(val, line)

    def _expr_stmt(self) -> ExprStmt:
        line = self._cur().line
        e = self._expr()
        self._eat(TT.SEMI)
        return ExprStmt(e, line)

    # ── expressions (precedence climbing) ────────────────────────────────────

    def _expr(self) -> Expr:
        return self._assign()

    def _assign(self) -> Expr:
        if self._at(TT.IDENT) and self._peek().ty == TT.EQ:
            line = self._cur().line
            name = self._advance().value
            self._advance()  # consume '='
            return AssignExpr(name, self._assign(), line)
        return self._or()

    def _or(self) -> Expr:
        left = self._and()
        while self._at(TT.OR):
            line = self._cur().line
            self._advance()
            left = BinOp("||", left, self._and(), line)
        return left

    def _and(self) -> Expr:
        left = self._eq()
        while self._at(TT.AND):
            line = self._cur().line
            self._advance()
            left = BinOp("&&", left, self._eq(), line)
        return left

    def _eq(self) -> Expr:
        left = self._cmp()
        while self._at(TT.EQ_EQ, TT.BANG_EQ):
            line = self._cur().line
            op = self._advance().value
            left = BinOp(op, left, self._cmp(), line)
        return left

    def _cmp(self) -> Expr:
        left = self._add()
        while self._at(TT.LT, TT.GT, TT.LT_EQ, TT.GT_EQ):
            line = self._cur().line
            op = self._advance().value
            left = BinOp(op, left, self._add(), line)
        return left

    def _add(self) -> Expr:
        left = self._mul()
        while self._at(TT.PLUS, TT.MINUS):
            line = self._cur().line
            op = self._advance().value
            left = BinOp(op, left, self._mul(), line)
        return left

    def _mul(self) -> Expr:
        left = self._unary()
        while self._at(TT.STAR, TT.SLASH):
            line = self._cur().line
            op = self._advance().value
            left = BinOp(op, left, self._unary(), line)
        return left

    def _unary(self) -> Expr:
        if self._at(TT.BANG):
            line = self._cur().line
            self._advance()
            return UnaryOp("!", self._unary(), line)
        if self._at(TT.MINUS):
            line = self._cur().line
            self._advance()
            return UnaryOp("-", self._unary(), line)
        return self._call()

    def _call(self) -> Expr:
        prim = self._primary()
        if isinstance(prim, IdentExpr) and self._at(TT.LPAREN):
            line = self._cur().line
            self._advance()
            args: list[Expr] = []
            if not self._at(TT.RPAREN):
                args.append(self._expr())
                while self._at(TT.COMMA):
                    self._advance()
                    args.append(self._expr())
            self._eat(TT.RPAREN)
            return CallExpr(prim.name, args, line)
        return prim

    def _primary(self) -> Expr:
        tok = self._cur()
        if tok.ty == TT.INT:
            self._advance(); return IntLit(int(tok.value), tok.line)
        if tok.ty == TT.TRUE:
            self._advance(); return BoolLit(True, tok.line)
        if tok.ty == TT.FALSE:
            self._advance(); return BoolLit(False, tok.line)
        if tok.ty == TT.IDENT:
            self._advance(); return IdentExpr(tok.value, tok.line)
        if tok.ty == TT.LPAREN:
            self._advance()
            e = self._expr()
            self._eat(TT.RPAREN)
            return e
        raise ParseError(
            f"Line {tok.line}: unexpected token {tok.ty.name} ({tok.value!r})"
        )


def parse(src: str) -> Program:
    return Parser(tokenize(src)).parse_program()

File: modules/compiler/rust/src/parser.rs
//! Recursive descent parser for Zap.

use crate::ast::*;
use crate::lexer::{Token, TT};

#[derive(Debug)]
pub struct ParseError(pub String);

impl std::fmt::Display for ParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub struct Parser {
    tokens: Vec<Token>,
    pos:    usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, pos: 0 }
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    fn cur(&self) -> &Token      { &self.tokens[self.pos] }
    fn peek(&self) -> &Token     { &self.tokens[(self.pos + 1).min(self.tokens.len() - 1)] }
    fn at(&self, ty: &TT) -> bool { &self.cur().ty == ty }
    fn line(&self) -> usize      { self.cur().line }

    fn advance(&mut self) -> Token {
        let tok = self.tokens[self.pos].clone();
        self.pos += 1;
        tok
    }

    fn eat(&mut self, ty: TT) -> Result<Token, ParseError> {
        let tok = self.cur().clone();
        if tok.ty != ty {
            return Err(ParseError(format!(
                "Line {}: expected {:?}, got {:?} ({:?})",
                tok.line, ty, tok.ty, tok.value
            )));
        }
        self.pos += 1;
        Ok(tok)
    }

    // ── top level ─────────────────────────────────────────────────────────────

    pub fn parse_program(&mut self) -> Result<Program, ParseError> {
        let mut decls = Vec::new();
        while !self.at(&TT::Eof) {
            if self.at(&TT::Fn) {
                decls.push(Decl::Fn(self.parse_fn()?));
            } else {
                decls.push(Decl::Stmt(self.parse_stmt()?));
            }
        }
        Ok(Program { decls })
    }

    fn parse_fn(&mut self) -> Result<FnDecl, ParseError> {
        let line = self.line();
        self.eat(TT::Fn)?;
        let name = self.eat(TT::Ident)?.value;
        self.eat(TT::LParen)?;
        let mut params = Vec::new();
        if !self.at(&TT::RParen) {
            params.push(self.parse_param()?);
            while self.at(&TT::Comma) {
                self.advance();
                params.push(self.parse_param()?);
            }
        }
        self.eat(TT::RParen)?;
        self.eat(TT::Arrow)?;
        let return_type = self.parse_type()?;
        let body = self.parse_block()?;
        Ok(FnDecl { name, params, return_type, body, line })
    }

    fn parse_param(&mut self) -> Result<(String, ZapType), ParseError> {
        let name = self.eat(TT::Ident)?.value;
        self.eat(TT::Colon)?;
        let ty = self.parse_type()?;
        Ok((name, ty))
    }

    fn parse_type(&mut self) -> Result<ZapType, ParseError> {
        let tok = self.cur().clone();
        match tok.ty {
            TT::TInt  => { self.advance(); Ok(ZapType::Int) }
            TT::TBool => { self.advance(); Ok(ZapType::Bool) }
            TT::TVoid => { self.advance(); Ok(ZapType::Void) }
            _ => Err(ParseError(format!("Line {}: expected type, got {:?}", tok.line, tok.value))),
        }
    }

    fn parse_block(&mut self) -> Result<Block, ParseError> {
        self.eat(TT::LBrace)?;
        let mut stmts = Vec::new();
        while !self.at(&TT::RBrace) && !self.at(&TT::Eof) {
            stmts.push(self.parse_stmt()?);
        }
        self.eat(TT::RBrace)?;
        Ok(Block { stmts })
    }

    // ── statements ────────────────────────────────────────────────────────────

    fn parse_stmt(&mut self) -> Result<Stmt, ParseError> {
        match self.cur().ty {
            TT::Let    => self.parse_let(),
            TT::Return => self.parse_return(),
            TT::If     => self.parse_if(),
            TT::While  => self.parse_while(),
            TT::Print  => self.parse_print(),
            _          => self.parse_expr_stmt(),
        }
    }

    fn parse_let(&mut self) -> Result<Stmt, ParseError> {
        let line = self.line();
        self.eat(TT::Let)?;
        let name = self.eat(TT::Ident)?.value;
        self.eat(TT::Colon)?;
        let ty = self.parse_type()?;
        self.eat(TT::Eq)?;
        let value = self.parse_expr()?;
        self.eat(TT::Semi)?;
        Ok(Stmt::Let(LetStmt { name, ty, value, line }))
    }

    fn parse_return(&mut self) -> Result<Stmt, ParseError> {
        let line = self.line();
        self.eat(TT::Return)?;
        let value = if self.at(&TT::Semi) { None } else { Some(self.parse_expr()?) };
        self.eat(TT::Semi)?;
        Ok(Stmt::Return(ReturnStmt { value, line }))
    }

    fn parse_if(&mut self) -> Result<Stmt, ParseError> {
        let line = self.line();
        self.eat(TT::If)?;
        let cond = self.parse_expr()?;
        let then_block = self.parse_block()?;
        let else_block = if self.at(&TT::Else) {
            self.advance();
            Some(self.parse_block()?)
        } else {
            None
        };
        Ok(Stmt::If(IfStmt { cond, then_block, else_block, line }))
    }

    fn parse_while(&mut self) -> Result<Stmt, ParseError> {
        let line = self.line();
        self.eat(TT::While)?;
        let cond = self.parse_expr()?;
        let body = self.parse_block()?;
        Ok(Stmt::While(WhileStmt { cond, body, line }))
    }

    fn parse_print(&mut self) -> Result<Stmt, ParseError> {
        let line = self.line();
        self.eat(TT::Print)?;
        self.eat(TT::LParen)?;
        let value = self.parse_expr()?;
        self.eat(TT::RParen)?;
        self.eat(TT::Semi)?;
        Ok(Stmt::Print(PrintStmt { value, line }))
    }

    fn parse_expr_stmt(&mut self) -> Result<Stmt, ParseError> {
        let line = self.line();
        let expr = self.parse_expr()?;
        self.eat(TT::Semi)?;
        Ok(Stmt::Expr(ExprStmt { expr, line }))
    }

    // ── expressions ───────────────────────────────────────────────────────────

    fn parse_expr(&mut self) -> Result<Expr, ParseError> {
        self.parse_assign()
    }

    fn parse_assign(&mut self) -> Result<Expr, ParseError> {
        if self.at(&TT::Ident) && self.peek().ty == TT::Eq {
            let line  = self.line();
            let name  = self.advance().value;
            self.advance(); // consume '='
            let value = Box::new(self.parse_assign()?);
            return Ok(Expr::Assign { name, value, line });
        }
        self.parse_or()
    }

    fn parse_or(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_and()?;
        while self.at(&TT::Or) {
            let line = self.line();
            self.advance();
            left = Expr::BinOp { op: "||".into(), left: Box::new(left), right: Box::new(self.parse_and()?), line };
        }
        Ok(left)
    }

    fn parse_and(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_eq()?;
        while self.at(&TT::And) {
            let line = self.line();
            self.advance();
            left = Expr::BinOp { op: "&&".into(), left: Box::new(left), right: Box::new(self.parse_eq()?), line };
        }
        Ok(left)
    }

    fn parse_eq(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_cmp()?;
        while matches!(self.cur().ty, TT::EqEq | TT::BangEq) {
            let line = self.line();
            let op = self.advance().value;
            left = Expr::BinOp { op, left: Box::new(left), right: Box::new(self.parse_cmp()?), line };
        }
        Ok(left)
    }

    fn parse_cmp(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_add()?;
        while matches!(self.cur().ty, TT::Lt | TT::Gt | TT::LtEq | TT::GtEq) {
            let line = self.line();
            let op = self.advance().value;
            left = Expr::BinOp { op, left: Box::new(left), right: Box::new(self.parse_add()?), line };
        }
        Ok(left)
    }

    fn parse_add(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_mul()?;
        while matches!(self.cur().ty, TT::Plus | TT::Minus) {
            let line = self.line();
            let op = self.advance().value;
            left = Expr::BinOp { op, left: Box::new(left), right: Box::new(self.parse_mul()?), line };
        }
        Ok(left)
    }

    fn parse_mul(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_unary()?;
        while matches!(self.cur().ty, TT::Star | TT::Slash) {
            let line = self.line();
            let op = self.advance().value;
            left = Expr::BinOp { op, left: Box::new(left), right: Box::new(self.parse_unary()?), line };
        }
        Ok(left)
    }

    fn parse_unary(&mut self) -> Result<Expr, ParseError> {
        if self.at(&TT::Bang) {
            let line = self.line(); self.advance();
            return Ok(Expr::UnaryOp { op: "!".into(), operand: Box::new(self.parse_unary()?), line });
        }
        if self.at(&TT::Minus) {
            let line = self.line(); self.advance();
            return Ok(Expr::UnaryOp { op: "-".into(), operand: Box::new(self.parse_unary()?), line });
        }
        self.parse_call()
    }

    fn parse_call(&mut self) -> Result<Expr, ParseError> {
        let prim = self.parse_primary()?;
        if let Expr::Ident(ref name, _) = prim {
            if self.at(&TT::LParen) {
                let line = self.line();
                let name = name.clone();
                self.advance();
                let mut args = Vec::new();
                if !self.at(&TT::RParen) {
                    args.push(self.parse_expr()?);
                    while self.at(&TT::Comma) {
                        self.advance();
                        args.push(self.parse_expr()?);
                    }
                }
                self.eat(TT::RParen)?;
                return Ok(Expr::Call { name, args, line });
            }
        }
        Ok(prim)
    }

    fn parse_primary(&mut self) -> Result<Expr, ParseError> {
        let tok = self.cur().clone();
        match tok.ty {
            TT::Int => {
                self.advance();
                let v: i64 = tok.value.parse().map_err(|_| {
                    ParseError(format!("Line {}: invalid integer {:?}", tok.line, tok.value))
                })?;
                Ok(Expr::IntLit(v, tok.line))
            }
            TT::True  => { self.advance(); Ok(Expr::BoolLit(true,  tok.line)) }
            TT::False => { self.advance(); Ok(Expr::BoolLit(false, tok.line)) }
            TT::Ident => { self.advance(); Ok(Expr::Ident(tok.value, tok.line)) }
            TT::LParen => {
                self.advance();
                let e = self.parse_expr()?;
                self.eat(TT::RParen)?;
                Ok(e)
            }
            _ => Err(ParseError(format!(
                "Line {}: unexpected token {:?} ({:?})",
                tok.line, tok.ty, tok.value
            ))),
        }
    }
}

pub fn parse(src: &str) -> Result<Program, String> {
    let tokens = crate::lexer::tokenize(src).map_err(|e| e.to_string())?;
    Parser::new(tokens).parse_program().map_err(|e| e.to_string())
}
